File: src/main/java/cz/solutions/cockroach/EsppReportPreparation.java
package cz.solutions.cockroach;

import lombok.Value;
import org.joda.time.LocalDate;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class EsppReportPreparation {
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormat.forPattern("dd.MM.YYYY").withZoneUTC();

    public EsppReport generateEsppReport(List<EsppRecord> esppRecordList,
                                         List<SaleRecord> saleRecordList,
                                         DateInterval interval,
                                         ExchangeRateProvider exchangeRateProvider) {
        List<EsppRecord> esppRecords = esppRecordList.stream()
                .filter(a -> interval.contains(a.getPurchaseDate()))
                .sorted(Comparator.comparing(EsppRecord::getDate))
                .toList();

        List<SaleRecord> esppSaleRecords = saleRecordList.stream()
                .filter(a -> interval.contains(a.getDate()))
                .filter(a->a.getType().equals("ESPP"))
                .sorted(Comparator.comparing(SaleRecord::getDate))
                .toList();


        ArrayList<PrintableEspp> printableEsppList = new ArrayList<>();
        double profitDolarValue = 0;
        double profitCroneValue = 0;
        double taxablePofitCroneValue = 0;
        int totalEsppAmount = 0;

        for (EsppRecord espp : esppRecords) {
            double soldQuantity = esppSaleRecords.stream()
                    .filter(it -> espp.getPurchaseDate().equals(it.getPurchaseDate()))
                    .mapToDouble(SaleRecord::getQuantity)
                    .sum();

            EsppInfo esppInfo = withConvertedPrices(espp,soldQuantity,exchangeRateProvider);


            printableEsppList.add(esppInfo.toPrintable());
            profitDolarValue += esppInfo.getBuyProfitValue();
            profitCroneValue += esppInfo.getBuyCronePofitValue();

            totalEsppAmount += espp.getQuantity();
            taxablePofitCroneValue+=esppInfo.getTaxableBuyCronePofitValue();
        }

        return new EsppReport(
                printableEsppList,
                profitCroneValue,
                profitDolarValue,
                totalEsppAmount,
                taxablePofitCroneValue
        );
    }

        private EsppInfo  withConvertedPrices(EsppRecord espp, double soldAmount, ExchangeRateProvider exchangeRateProvider){
            double exchange = exchangeRateProvider.rateAt(espp.getPurchaseDate());
            double partialProfit = espp.getPurchaseFmv() - espp.getPurchasePrice();

            return new EsppInfo(
                    espp.getPurchaseDate(),
                    espp.getQuantity(),
                    exchange,
                    espp.getPurchasePrice(),
                    espp.getPurchaseFmv(),
                    partialProfit,
                    partialProfit * espp.getQuantity(),
                    partialProfit * espp.getQuantity() * exchange,
                    soldAmount,
                    partialProfit*soldAmount*exchange
            );
        }

        @Value
        private static class EsppInfo {
            LocalDate date;
            int amount;
            double exchange;
            double onePricePurchaseDolarValue;
            double onePriceDolarValue;
            double oneProfitValue;
            double buyProfitValue;
            double buyCronePofitValue;
            double soldAmount;
            double taxableBuyCronePofitValue;

            PrintableEspp toPrintable() {
                return new PrintableEspp(
                        DATE_FORMATTER.print(date),
                        amount,
                        FormatingHelper.formatExchangeRate(exchange),
                        FormatingHelper.formatDouble(onePricePurchaseDolarValue),
                        FormatingHelper.formatDouble(onePriceDolarValue),
                        FormatingHelper.formatDouble(oneProfitValue),
                        FormatingHelper.formatDouble(buyProfitValue),
                        FormatingHelper.formatDouble(buyCronePofitValue),
                        soldAmount,
                        FormatingHelper.formatDouble(taxableBuyCronePofitValue)
                );
            }
        }

}

File: src/main/kotlin/cz/solutions/cockroach/EsppReportPreparation.kt
package cz.solutions.cockroach

import org.joda.time.LocalDate
import org.joda.time.format.DateTimeFormat
import org.joda.time.format.DateTimeFormatter

object EsppReportPreparation {
    private val DATE_FORMATTER: DateTimeFormatter = DateTimeFormat.forPattern("dd.MM.YYYY").withZoneUTC()

    fun generateEsppReport(
        esppRecordList: List<EsppRecord>,
        saleRecordList: List<SaleRecord>,
        interval: DateInterval,
        exchangeRateProvider: ExchangeRateProvider,
        taxableAmount: (Double, Double) -> Double = { quantity,_ -> quantity }
    ): EsppReport {
        val esppRecords = esppRecordList
            .filter { interval.contains(it.purchaseDate) }
            .sortedBy { it.date }

        val esppSaleRecordsByDate = saleRecordList
            .filter { interval.contains(it.date) && it.type == "ESPP" }
            .groupBy { it.purchaseDate }

        val esppInfos = esppRecords.map { espp ->
            val soldQuantity = esppSaleRecordsByDate[espp.purchaseDate]
                ?.sumOf { it.quantity } ?: 0.0
            withConvertedPrices(espp, soldQuantity,taxableAmount(espp.quantity,soldQuantity), exchangeRateProvider)
        }

        return EsppReport(
            printableEsppList = esppInfos.map { it.toPrintable() },
            profitCroneValue = esppInfos.sumOf { it.buyCroneProfitValue },
            profitDolarValue = esppInfos.sumOf { it.buyProfitValue },
            totalEsppAmount = esppInfos.sumOf { it.amount },
            taxableProfitCroneValue = esppInfos.sumOf { it.taxableBuyCroneProfitValue }
        )
    }

    private fun withConvertedPrices(espp: EsppRecord, soldAmount: Double, taxableAmount: Double, exchangeRateProvider: ExchangeRateProvider): EsppInfo {
        val exchange = exchangeRateProvider.rateAt(espp.purchaseDate, Currency.USD)
        val partialProfit = espp.purchaseFmv - espp.purchasePrice

        return EsppInfo(
            symbol = espp.symbol,
            broker = espp.broker,
            date = espp.purchaseDate,
            amount = espp.quantity,
            exchange = exchange,
            onePricePurchaseDolarValue = espp.purchasePrice,
            onePriceDolarValue = espp.purchaseFmv,
            oneProfitValue = partialProfit,
            buyProfitValue = partialProfit * espp.quantity,
            buyCroneProfitValue = partialProfit * espp.quantity * exchange,
            soldAmount = soldAmount,
            taxableBuyCroneProfitValue = partialProfit * taxableAmount * exchange
        )
    }

    private data class EsppInfo(
        val symbol: String,
        val broker: String,
        val date: LocalDate,
        val amount: Double,
        val exchange: Double,
        val onePricePurchaseDolarValue: Double,
        val onePriceDolarValue: Double,
        val oneProfitValue: Double,
        val buyProfitValue: Double,
        val buyCroneProfitValue: Double,
        val soldAmount: Double,
        val taxableBuyCroneProfitValue: Double
    ) {
        fun toPrintable(): PrintableEspp {
            return PrintableEspp(
                symbol = symbol,
                broker = broker,
                date = DATE_FORMATTER.print(date),
                amount = amount,
                exchange = FormatingHelper.formatExchangeRate(exchange),
                onePricePurchaseDolarValue = FormatingHelper.formatDouble(onePricePurchaseDolarValue),
                onePriceDolarValue = FormatingHelper.formatDouble(onePriceDolarValue),
                oneProfitValue = FormatingHelper.formatDouble(oneProfitValue),
                buyProfitValue = FormatingHelper.formatDouble(buyProfitValue),
                buyCroneProfitValue = FormatingHelper.formatDouble(buyCroneProfitValue),
                soldAmount = soldAmount,
                taxableBuyCroneProfitValue = FormatingHelper.formatDouble(taxableBuyCroneProfitValue)
            )
        }
    }
}